CMakeLists.txt

```diff
@@ -36,6 +36,8 @@ option(IREE_ENABLE_COMPILER_TRACING "Enables instrumented compiler tracing." OFF
 option(IREE_ENABLE_THREADING "Builds IREE in with thread library support." ON)
 option(IREE_ENABLE_CLANG_TIDY "Builds IREE in with clang tidy enabled on IREE's libraries." OFF)
 
+option(IREE_ENABLE_MACRO_FOR_EXTERNAL_PROJECT "Use IREE macros to build objects in external projects." OFF)
```

Comment:
Also, the macros should really an internal implementation detail, so I hesitate to endorse their use in downstream projects :/

Or, to rephrase that: it must be possible to use IREE in downstream projects without using IREE's helper functions - in CMake or with some other build system - particularly for the runtime. I don't want downstream users to feel that they have to use IREE's build system functions.

That does get tricky when you want to run the compiler under the build system though. It's easier to compile offline and then reference the compiled artifacts as part of your runtime-only build. Thinking from a game engine or mobile app perspective, you could consider ML programs to be like images, shaders, audio files, and other assets. While you certainly can hook up pngcrush and all sorts of tools as part of your build system, that isn't a hard requirement.

Reply:
Yeah I really don't think we should be endorsing usage of these in external projects. They are not designed or maintained to be reusable components. If they were, they should live outside of the IREE repository. I'm fine with improvements that make them more usable in general, but their usage outside of the IREE project should be considered totally unsupported IMO. I'm opposed to the addition of an option like this on those grounds.

Reply:
Totally agree building IREE runtime should not _rely_ on these macros, but these macros do make the development much easier. Instead of setting target dependencies and configurations with 20+ lines per target (as in https://github.com/google/iree-samples/blob/main/cpp/vision_inference/CMakeLists.txt), it is really beneficial to just use the macros. I understand the usage is not warranted, but it does provide a cleaner build layout in the CMakeLists.txt.

May I propose the following: Add an optional argument in `iree_cc_library` and such to define the namespace, while passthrough the namespace in `iree_package_ns` once it is set. Do you think that can be an acceptable solution?

Reply:
I think adding a branch for files not located under `${IREE_ROOT_DIR}` would be minimally invasive. It would allow downstream projects to use the macros, but would not offer too much configuration control. More serious uses in downstream projects should create their own functions tailored to their specific usage.

Reply:
Yes, that's exactly my approach now. :)